### PR TITLE
Implementation of Natural Numbers 

### PR DESCRIPTION
**Pull request by agureev**
1) Introduces SeqN category of finite natural number sequences

2) Introduces a new pipeline compiling STLC terms through SeqN instead of BITC

3) Implements for each n an object of natural numbers of n-bit length alongside appropriate floored operations alongside new API. Propagated the implementation to STLC and SeqN.

4) Implements is-less-than and is-equal-to predicates on natural numbers in Geb, propagates the change to STLC and SeqN.

5) Upgrades the pipeline now supporting compilation of STLC+Natural numbers+Natural Number Predicates to Vamp-IR

### CODE REVIEW

File: src/seqn/seqn.lisp [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+(defmethod seqwidth ((obj <substobj>))
+  (typecase-of substobj obj
+    (so0 (list 0))
+    (so1 (list 0))
+    (coprod (cons 1 (mapcar #'max-list
+                            (seq-max-fill
+                             (seqwidth
+                              (mcar obj))
+                             (seqwidth
+                              (mcadr obj))))))
```
Comment (mariari) [inline src/seqn/seqn.lisp:53]: reading this makes it a bit hard to read, formatting it like this is more obvious to my eyes at least

```lisp
    (coprod (cons 1
                  (mapcar #'max-list
                          (seq-max-fill (seqwidth (mcar obj))
                                        (seqwidth (mcadr obj))))))
```

File: src/seqn/seqn.lisp [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+  "Takes two lists, makes them same length by adding 0es on the right
+where necessary and takes their pointwise product"
+  (let ((l1 (length seq1))
+        (l2 (length seq2)))
+    (if (< l1 l2)
+        (prod-list (fill-in (- l2 l1) seq1) seq2)
+        (prod-list seq1 (fill-in (- l1 l2) seq2)))))
+
+(defun max-list (lst)
+  (max (car lst) (car (cdr lst))))
```
[the remark this answers is not in the document]
Comment (mariari) [reply src/seqn/seqn.lisp:42]: Further the name is misleading, wouldn't you want this?

```lisp
(defun max-list (lst)
  (reduce #'max lst :initial-value 0))
```

File: src/seqn/seqn.lisp [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+                              (mcar obj))
+                             (seqwidth
+                              (mcadr obj))))))
+    (prod (append (seqwidth (mcar obj))
+                  (seqwidth (mcadr obj))))
+    (otherwise (geb.utils:subclass-responsibility obj))))
+
+(defun inj-coprod-parallel (
+                            obj     ;already transformed mcar or mcadr
+                            copr    ;already transformed coproduct
```
Comment (mariari) [inline src/seqn/seqn.lisp:65]: No. Don't comment like this. Just add it in the documentation string so it's reported to anyone who wants to use this

File: src/seqn/seqn.lisp [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+                            )
+  "takes an A x B and gives an appropriate injection of
+(a1,...,an) into (max (a1, b1), ...., max(an, bn),...)
+i.e. where the maxes are being taken during the seqwidth operation
+without fulling in of the smaller object"
+  (if (= (length obj) (- (length copr) 1))
+      (drop-width obj  (cdr copr))
+      (composition (drop-width (append obj
+                                       (0-list (- (1- (length copr))
+                                                  (length obj))))
```
Comment (mariari) [inline src/seqn/seqn.lisp:75]: Drop the append, and do something like I suggested earlier

```diff
+ (pad-with-zeros obj (1- (length copr))
- (append obj (0-list ...))
``

File: src/seqn/seqn.lisp [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+(defmethod dom ((x <seqn>))
+  "Gives the domain of a morphism in SeqN.
+For a less formal desription consult the specs file"
+  (typecase-of seqn x
+    (composition      (dom (mcadr x)))
+    (fork-seq         (mcar x))
+    (parallel-seq     (append (dom (mcar x)) (dom (mcadr x))))
+    (id               (mcar x))
+    (drop-nil         (mcar x))
+    (remove-right     (append (mcar x) (list 0)))
```
Comment (mariari) [inline src/seqn/seqn.lisp:89]: so many appens on stuff, maybe we should consider a different data structure at some point.

File: src/seqn/trans.lisp [hunk 6]
```diff
@@ -0,0 +1,399 @@
+(in-package :geb.seqn.trans)
+
+(defun remove0-dom (morphism)
+  (remove 0 (dom morphism)))
+
+;; Add range-n function
+
+(defun range-constraints-dom (domain)
+  (cond ((zerop (length domain))   nil)
+        ((= (length domain) 1)    (list (geb.vampir:range-n
```
Comment (mariari) [inline src/seqn/trans.lisp:10]:
```lisp
MAIN> (=  (length '(1)) 1)
T
MAIN> (not (cdr '(1)))
T
MAIN> (= (length '(1 2)) 1)
NIL
MAIN> (not (cdr '(1 2)))
NIL
MAIN> 
```

this is a bit cheaper and achieves the same result

File: src/seqn/trans.lisp [hunk 7]
```diff
@@ -0,0 +1,399 @@
...
+                                         (vamp:make-constant :const (car domain))
+                                         (vamp:make-wire :var
+                                                         (intern (format nil "x~a" 1))))))
+        (t (cons (geb.vampir:range-n (vamp:make-constant :const (car (last domain)))
+                                (vamp:make-wire :var
+                                                (intern (format nil "x~a" (length domain)))))
+            (range-constraints-dom (butlast
+                                    domain))))))
+
```
Comment (mariari) [inline src/seqn/trans.lisp:19]: formatting note, just newline the cond answers at this point

```lisp
(defun range-constraints-dom (domain)
  (cond ((zerop (length domain))
         nil)
        ((= (length domain) 1)
         (list (geb.vampir:range-n
                (vamp:make-constant :const (car domain))
                (vamp:make-wire :var
                                (intern (format nil "x~a" 1))))))
        (t
         (cons (geb.vampir:range-n
                (vamp:make-constant :const (car (last domain)))
                (vamp:make-wire :var
                                (intern (format nil "x~a" (length domain)))))
               (range-constraints-dom (butlast domain))))))
```

much easier to read. See my comments on this to see what else can be changed. This does not include the proper code changes.

File: src/seqn/trans.lisp [hunk 8]
```diff
@@ -0,0 +1,399 @@
+(in-package :geb.seqn.trans)
+
+(defun remove0-dom (morphism)
+  (remove 0 (dom morphism)))
+
+;; Add range-n function
+
+(defun range-constraints-dom (domain)
+  (cond ((zerop (length domain))   nil)
```
Comment (mariari) [inline src/seqn/trans.lisp:9]: Just use `(null domain)`. Much cheaper and easier to understand

File: src/seqn/trans.lisp [hunk 9]
```diff
@@ -0,0 +1,399 @@
...
+                                    domain))))))
+
+(defmethod to-circuit ((morphism <seqn>) name)
+  "Turns a SeqN term into a Vamp-IR Gate with the given name
+Note that what is happening is that we look at the domain of the morphism
+and skip 0es, making non-zero entries into wires"
+  (let* ((wire-count (length (dom morphism)))
+         (wires (loop for i from 1 to wire-count
+                      collect (vamp:make-wire :var (intern (format nil "x~a" i)
+                                                           :keyword)))))
```
Comment (mariari) [inline src/seqn/trans.lisp:27]:
```lisp
MAIN> (mapcar (lambda (i) (intern (format nil "X~a" i) :keyword)) (alexandria:iota 10 :start 1))
(:X1 :X2 :X3 :X4 :X5 :X6 :X7 :X8 :X9 :X10)
```

because you are using `intern` using `"x"` and not `"X"` makes a weird symbol that my vmapir printer already fixes

File: src/seqn/trans.lisp [hunk 10]
```diff
@@ -0,0 +1,399 @@
...
+                                     (list
+                                      (vamp:make-tuples :wires
+                                                        (remove nil (mapcar (lambda (x) (if (= (car x) 0)
+                                                                                            nil
+                                                                                            (cadr x)))
+                                                                            (prod-list (cod morphism)
+                                                                                       (to-vampir morphism
+                                                                                                  wires
+                                                                                                  nil))))
+                                                        )))))))
```
Comment (mariari) [inline src/seqn/trans.lisp:41]: Bad formatting, let me fix that first, before logic.

```lisp
(defmethod to-circuit ((morphism <seqn>) name)
  "Turns a SeqN term into a Vamp-IR Gate with the given name
Note that what is happening is that we look at the domain of the morphism
and skip 0es, making non-zero entries into wires"
  (let* ((wire-count (length (dom morphism)))
         (wires (loop for i from 1 to wire-count
                      collect (vamp:make-wire :var (intern (format nil "x~a" i)
                                                           :keyword)))))
    (list
     (vamp:make-alias
      :name name
      :inputs wires
      :body  (append
              (range-constraints-dom (dom morphism))
              (list
               (vamp:make-tuples
                :wires
                (remove nil
                        (mapcar (lambda (x) (if (= (car x) 0)
                                            nil
                                            (cadr x)))
                                (prod-list (cod morphism)
                                           (to-vampir morphism wires nil)))))))))))
```

Comment (mariari) [reply src/seqn/trans.lisp:41]: Now time for the logic changes.

Please don't do a filter map with remove and mapcar. we can be more clever.

Here are the change that you should make
```lisp
MAIN> (mapcan (lambda (x) (if (zerop (car x)) nil (list (cadr x))))
              '((0 1) (2 3) (4 5)))
(3 5)
MAIN> (filter-map (lambda (x) (if (zerop (car x)) nil (cadr x)))
                  '((0 1) (2 3) (4 5)))
(3 5)
MAIN> 
```

I.E. use filter-map or mapcan, and use `zerop` over = 0

File: src/seqn/trans.lisp [hunk 11]
```diff
@@ -0,0 +1,399 @@
...
+(defmethod to-vampir ((obj inj-length-left) inputs constraints)
+  "Look at the MCAR. Look at non-null wide entries and place
+0-es in the outputs otherwise ignore"
+  (declare (ignore constraints))
+  (let ((l (length  (mcadr obj))))
+    (if (zerop l)
+        inputs
+        (append inputs (mapcar (lambda (x)
+                          (vamp:make-constant :const x))
+                               (0-list l))))))
```
Comment (mariari) [inline src/seqn/trans.lisp:125]: what is this formatting, plz fix

Comment (mariari) [reply src/seqn/trans.lisp:125]: also for the list, just do something like this

```lisp
(make-list l :initial-element (vamp:make-constant :const 0))
```

File: src/seqn/trans.lisp [hunk 12]
```diff
@@ -0,0 +1,399 @@
...
+  "Look at the MCADR. Look at non-null wide entries and place
+0-es in the outputs "
+  (declare (ignore constraints))
+  (let ((l (length  (mcar obj))))
+    (if (zerop l)
+        inputs
+        (append (mapcar (lambda (x)
+                          (vamp:make-constant :const x))
+                        (0-list l))
+                inputs))))
```
Comment (mariari) [inline src/seqn/trans.lisp:137]: Same here

File: src/seqn/trans.lisp [hunk 13]
```diff
@@ -0,0 +1,399 @@
...
+        (append inputs (mapcar (lambda (x)
+                          (vamp:make-constant :const x))
+                               (0-list l))))))
+
+(defmethod to-vampir ((obj inj-length-right) inputs constraints)
+  "Look at the MCADR. Look at non-null wide entries and place
+0-es in the outputs "
+  (declare (ignore constraints))
+  (let ((l (length  (mcar obj))))
+    (if (zerop l)
```
Comment (mariari) [inline src/seqn/trans.lisp:132]: Can probably cut the length check doing null for the if and using length in the else branch for make-list

File: src/seqn/trans.lisp [hunk 14]
```diff
@@ -0,0 +1,399 @@
...
+                                                                        carobj))
+                                      ((= cadrobjnum 1) (infix-creation :+
+                                                                        (infix-creation :*
+                                                                                        ifzero
+                                                                                        carobj)
+                                                                        car))
+                                      (t  output))))
+                             (t output))))
+         (prod-list left
+                    right)))))
```
Comment (mariari) [inline src/seqn/trans.lisp:219]: This looks big, can you abstract it at all? I am seeing some repeat patterns, but can we somehow make this cleaner? I should be able to at a glance figure out some things, this is a bit intimidating

File: src/seqn/trans.lisp [hunk 15]
```diff
@@ -0,0 +1,399 @@
...
+
+(defmethod to-vampir ((obj shift-front) inputs constraints)
+  "Takes the MCADR entry and moves it upward leaving everything
+else fixed. Note that we have to be careful as inputs will have 0es
+removed already and hence we cannot count as usual"
+  (declare (ignore constraints))
+  (append (cons (nth (1- (mcadr obj))
+                     inputs)
+                (subseq inputs 0 (1- (mcadr obj))))
+          (subseq inputs (mcadr obj))))
```
Comment (mariari) [inline src/seqn/trans.lisp:229]: let the mcadr as you use it 3 times.

also for uniformity reasons it might be good to do 1 append, with the following
```lisp
(append (list moved-element) initial rest)
```

would make the code obvious. Here I had to read it through a bit to catch it.

File: src/seqn/trans.lisp [hunk 16]
```diff
@@ -0,0 +1,399 @@
...
+    (if (and (typep car 'vamp:constant)
+             (typep cadr 'vamp:constant))
+        (let ((plus (+ (vamp:const car)
+                       (vamp:const cadr))))
+          (if (> (expt 2 mcar) plus)
+              (vamp:make-constant :const plus)
+              (error "Range Exceeded")))
+        (list (geb.vampir:plus-range (vamp:make-constant :const mcar)
+                                     car
+                                     cadr)))))
```
Comment (mariari) [inline src/seqn/trans.lisp:253]: we should probably alias vampir better

File: src/seqn/trans.lisp [hunk 17]
```diff
@@ -0,0 +1,399 @@
...
+
+(defmethod to-vampir ((obj seqn-subtract) inputs constraints)
+  (declare (ignore constraints))
+  (let ((car (car inputs))
+        (cadr (cadr inputs)))
+    (if (and (typep car 'vamp:constant)
+             (typep cadr 'vamp:constant))
```
Comment (mariari) [inline src/seqn/trans.lisp:260]: This pattern is very common, should we make a function that checks if the first two are constants?

File: src/seqn/trans.lisp [hunk 18]
```diff
@@ -0,0 +1,399 @@
...
+                                      cadr)))))
+
+(defmethod to-vampir ((obj seqn-multiply) inputs constraints)
+  (declare (ignore constraints))
+  (let ((car (car inputs))
+        (cadr (cadr inputs))
+        (mcar (mcar obj)))
+    (if (and (typep car 'vamp:constant)
+             (typep cadr 'vamp:constant))
+        (let ((mult (* (vamp:const car) (vamp:const cadr))))
```
Comment (mariari) [inline src/seqn/trans.lisp:276]: shows up here as well..

I think maybe we should have a stack like operation so we can just apply * - etc on it

note we can use `destructure-bind` to potentially make this easier to write

http://clhs.lisp.se/Body/m_destru.htm

File: src/seqn/trans.lisp [hunk 19]
```diff
@@ -0,0 +1,399 @@
...
+        (list (geb.vampir:mult-range (vamp:make-constant :const mcar)
+                                     car
+                                     cadr)))))
+
+(defmethod to-vampir ((obj seqn-divide) inputs constraints)
+  (declare (ignore constraints))
+  (let ((car (car inputs))
+        (cadr (cadr inputs)))
+    (if (and (typep car 'vamp:constant)
+             (typep cadr 'vamp:constant))
```
Comment (mariari) [inline src/seqn/trans.lisp:289]: More of the same boilerplate

File: src/seqn/trans.lisp [hunk 20]
```diff
@@ -0,0 +1,399 @@
...
+                                         :arguments (list (vamp:make-constant :const (1- mcar))
+                                                          (geb.utils:apply-n dpth
+                                                                             (lambda (x)
+                                                                               (cadr
+                                                                                (vamp:arguments
+                                                                                 (car
+                                                                                  (vamp:arguments
+                                                                                   (cadr
+                                                                                    (vamp:arguments x)))))))
+                                                                             rng)))
```
Comment (mariari) [inline src/seqn/trans.lisp:355]: please fix formatting. you made the `:func` come up first, so by the time we get to `:arguments` it's already too indented for the rest of the logic. so the lambda looks like a mess, rather then a more natural formulation

File: src/seqn/trans.lisp [hunk 21]
```diff
@@ -0,0 +1,399 @@
...
+            (list (vamp:make-application :func :n_th
+                                         :arguments (list (vamp:make-constant :const (1- mcar))
+                                                          (geb.utils:apply-n dpth
+                                                                             (lambda (x)
+                                                                               (cadr
+                                                                                (vamp:arguments
+                                                                                 (car
+                                                                                  (vamp:arguments
+                                                                                   (cadr
+                                                                                    (vamp:arguments x)))))))
```
Comment (mariari) [inline src/seqn/trans.lisp:354]: What is this lambda, why are we cadrs and caring idk even know what this semantically is suppposed to be doing.

Very confusing, please refactor.

File: src/seqn/trans.lisp [hunk 22]
```diff
@@ -0,0 +1,399 @@
...
+                                                                             rng)))
+                  (geb.vampir:combine
+                   (vamp:make-application :func :drop_ith
+                                          :arguments lst)))))))
+
+(defun range-depth (x)
+  (let ((cadr (cadr (vamp:arguments x))))
+    (if (not (typep cadr 'vamp:application))
+        0
+        (1+ (range-depth (cadr (vamp:arguments (car (vamp:arguments cadr)))))))))
```
Comment (mariari) [inline src/seqn/trans.lisp:364]: you need to get rid of this argument pattern, make it have a reasonable name, as I can't make heads or tails of it.

File: src/seqn/trans.lisp [hunk 23]
```diff
@@ -0,0 +1,399 @@
...
+
+(defmethod to-vampir ((obj seqn-eq) inputs constraints)
+  (declare (ignore constraints))
+  (let ((car (car inputs))
+        (cadr (cadr inputs))
+        (zero (vamp:make-constant :const 0)))
+    (if (and (typep car 'vamp:constant)
+             (typep cadr 'vamp:constant))
+        (if (= 0 (- (vamp:const car) (vamp:const cadr)))
```
Comment (mariari) [inline src/seqn/trans.lisp:373]: use `zerop`

File: src/seqn/trans.lisp [hunk 24]
```diff
@@ -0,0 +1,399 @@
...
+                                                                                            (cadr x)))
+                                                                            (prod-list (cod morphism)
+                                                                                       (to-vampir morphism
+                                                                                                  wires
+                                                                                                  nil))))
+                                                        )))))))
+
+;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+;; SeqN to Vamp-IR Compilation
+;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
```
Comment (mariari) [inline src/seqn/trans.lisp:45]: General comment on Vampir @agureev :

Maybe we could remove the constant checks into an optimizable check, and have that logic in it's own check.

We could maybe even run it as a pass before this code.

As this code mixes generation with some optimziation, making both read messy. This can be done as a pass or by just doing a check and dispatching so at least it's in a separate place.

File: src/specs/package.lisp [hunk 25]
```diff
@@ -15,10 +15,14 @@
...
 
+(muffle-package-variance
+ (defpackage #:geb.seqn.spec
+   (:use #:geb.utils #:cl #:geb.mixins))) 
+
 ;; please document this later.
 (muffle-package-variance
  (uiop:define-package #:geb.lambda.spec
-   (:documentation "Basic spec for creating lambda terms")
+     (:documentation "Basic spec for creating lambda terms")
```
Comment (mariari) [inline src/specs/package.lisp:25]: remove this change

File: src/specs/seqn-printer.lisp [hunk 26]
```diff
@@ -0,0 +1,32 @@
+(in-package #:geb.seqn.spec)
+
+(defmacro easy-printer (class-name)
+  `(defmethod print-object ((obj ,class-name) stream)
+     (format stream "~A"
+             (cons ',class-name
+                   (mapcar #'cdr (geb.mixins:to-pointwise-list obj))))))
```
Comment (mariari) [inline src/specs/seqn-printer.lisp:7]: If we are copy and pasting this, maybe we should make a separate topic that makes this in the utils or something.

File: src/geb/geb.lisp [hunk 27]
```diff
@@ -11,6 +11,22 @@
...
+    (if (= num 1)
+        (pair (curry (comp fun
+                           (prod-mor fst
+                                     (nat-const 1 0))))
+              ;; On the left, see what the morphism does at 0
+              (curry (comp fun
+                           (prod-mor fst
+                                     (nat-const 1 1))))
+              ;; On the right, see what it does on 1
+              )
```
Comment (mariari) [inline src/geb/geb.lisp:25]: )'s are sad when they don't hug

File: src/geb/geb.lisp [hunk 28]
```diff
@@ -29,6 +45,19 @@
...
        (subclass-responsibility x)))))
 
+(defmethod dom ((x <natmorph>))
+  (cond ((typep x 'nat-concat)      (prod (nat-width (num-left x))
+                                          (nat-width (num-right x))))
+        ((typep x 'nat-const)       so1)
+        ((typep x 'nat-inj)         (nat-width (num x)))
+        ((typep x 'one-bit-to-bool) (nat-width 1))
+        ((typep x 'nat-decompose)   (nat-width (num x)))
+        (t                         (let ((bit (nat-width (num x))))
```
Comment (mariari) [inline src/geb/geb.lisp:55]: probably need 1 extra space here to match the rest of the formatting

File: src/geb/geb.lisp [hunk 29]
```diff
@@ -130,7 +181,7 @@ u
                morph dom)
         (comp morph (commutes (mcadr dom) (mcar dom))))))
 
-(-> !-> (substobj substobj) substobj)
```
Comment (mariari) [inline src/geb/geb.lisp:133]: uhhhhhh, this feels wrong, given it literally matches `a` if anything we may need to generalize `b` to `catsubstobj`.

File: src/geb/geb.lisp [hunk 30]
```diff
@@ -156,32 +207,59 @@ u
     (geb:so1          1)))
 
 
-(-> so-eval (substobj substobj) substmorph)
-(defun so-eval (x y)
+(-> so-eval (t t) substmorph)
+(defmethod so-eval ((x <substobj>) y)
```
Comment (mariari) [inline src/geb/geb.lisp:211]: we should remove the entire -> here, if it's not a method

File: src/geb/geb.lisp [hunk 31]
```diff
@@ -267,7 +345,7 @@ This is the morphism part of the product functor Geb x Geb → Geb"
   "Id")
 
 (defmethod maybe ((obj <substobj>))
-  "I recursively add maybe terms to all [\\<SBUSTOBJ\\>][class] terms,
+  "I recursively add maybe terms to all [\\<SUBSTOBJ\\>][class] terms,
```
Comment (mariari) [inline src/geb/geb.lisp:348]: I don't se the diff here, did an encoding of the word get better, did I fat finger a unicode character here by accident?

File: src/geb/geb.lisp [hunk 32]
```diff
@@ -329,7 +410,7 @@ codomain respectively"
 (defmethod gapply ((morph <substmorph>) object)
   "My main documentation can be found on [GAPPLY][generic-function]
 
-I am the [GAPPLY][generic-function] for [\\<SBUSTMORPH\\>][class], the
+I am the [GAPPLY][generic-function] for [\\<SUBSTMORPH\\>][class], the
```
Comment (mariari) [inline src/geb/geb.lisp:413]: same here

File: src/geb/package.lisp [hunk 33]
```diff
@@ -21,8 +21,12 @@
   (commutes          pax:function)
   (commutes-left     pax:function)
   (!->               pax:function)
-  (so-eval           pax:function)
-  (so-hom-obj        pax:function)
+  (so-eval           (pax:method () (<natobj>)))
+  (so-eval           (pax:method () (<substobj>)))
+  (so-eval           pax:generic-function)
```
Comment (mariari) [inline src/geb/package.lisp:26]: we should probably move this to the generic package we have for generic functions?

File: src/geb/trans.lisp [hunk 34]
```diff
@@ -189,3 +189,178 @@ should calculate
...
+                                                                               (length (seqwidth rt))))
+                                                                           (seqwidth rt))
+                                                               (drop-nil (nthcdr (1+ (length (seqwidth rt)))
+                                                                                 (seqwidth cpr))))))))
+          ((= (length (seqwidth rt))
+              (length (seqwidth lft)))
+           (branch-seq (composition (to-seqn (mcar obj))
+                                    (drop-width (cdr (seqwidth cpr)) (seqwidth lft)))
+                       (composition (to-seqn (mcadr obj))
+                                    (drop-width (cdr (seqwidth cpr)) (seqwidth rt))))))))
```
Comment (mariari) [inline src/geb/trans.lisp:292]: Abstract this, I can't read this just by looking. I'm sure you can find similarities and remove it. Too much code.

File: src/geb/trans.lisp [hunk 35]
```diff
@@ -189,3 +189,178 @@ should calculate
...
+
+(defmethod to-seqn ((obj project-left))
+  "Takes a list of an even length and cuts the right half"
+  (let ((sw (seqwidth (mcar obj))))
+    (composition (remove-right sw)
+                 (parallel-seq (id sw)
+                               (drop-nil (seqwidth (mcadr obj)))))))
```
Comment (mariari) [inline src/geb/trans.lisp:299]: why are we ending up with nils anyways. Seems like sloppy coding? Or does it have a point?

File: src/geb/trans.lisp [hunk 36]
```diff
@@ -189,3 +189,178 @@ should calculate
...
+
+(defmethod to-seqn  ((obj project-right))
+  "Takes a list of an even length and cuts the left half"
+  (let ((sw (seqwidth (mcadr obj))))
+    (composition (remove-left sw)
+                 (parallel-seq (drop-nil (seqwidth (mcar obj)))
```
Comment (mariari) [inline src/geb/trans.lisp:305]: same here

File: src/generics/package.lisp [hunk 37]
```diff
@@ -12,10 +12,11 @@ You can view their documentation in their respective API sections.
...
-  (to-cat     pax:generic-function)
-  (to-vampir  pax:generic-function))
+  (gapply         pax:generic-function)
+  (maybe          pax:generic-function)
+  (to-circuit     pax:generic-function)
+  (to-bitc        pax:generic-function)
+  (to-seqn        pax:generic-function)
+  (to-poly        pax:generic-function)
+  (to-cat         pax:generic-function)
+  (to-vampir      pax:generic-function))
```
Comment (mariari) [inline src/generics/package.lisp]: we should maybe move this to its own commit or if this is due to a new addition to that commit.

File: src/lambda/trans.lisp [hunk 38]
```diff
@@ -221,6 +233,99 @@ This follows from the fact that bool arapped in maybe is 1 + (bool + bool)"
...
+                                                            so1
+                                                            (ttype ltm))
+                                                (geb:pair (rec context ltm)
+                                                          (rec context rtm))))))
+               ((bit-choice bitv)  (comp (comp (->right so1 (ttype tterm))
+                                              (nat-const (ttype tterm) (reduce
+                                                                        #'(lambda (a b)
+                                                                            (+ (ash a 1) b))
+                                                                        bitv)))
+                                        (terminal (stlc-ctx-maybe context))))
```
Comment (mariari) [inline src/lambda/trans.lisp:328]: I see lambda is still horrible as ever.

Let us attept to clean it up.

Please fix the formatting by newlining after `(bit-choice bitv)`, so the code it maps to is on the next line, given just how big they are.

File: src/seqn/seqn.lisp [hunk 39]
```diff
@@ -141,4 +141,57 @@ For a less formal description consult the specs file"
...
+    (fork-seq         (append vector vector))
+    (drop-nil         (list 0))
+    (remove-right     (butlast vector))
+    (remove-left      (cdr vector))
+    (drop-width       vector)
+    (inj-length-left  (append vector
+                              (mapcar (lambda (x) 0)
+                                      (mcadr morphism))))
+    (inj-length-right (append (mapcar (lambda (x) 0)
+                                      (mcar morphism))
```
Comment (mariari) [inline src/seqn/seqn.lisp:167]: sbcl will yell at you for unused x. just use the `make-list` I used in my previous comments for these. would read much better. Or in this case use the `pad-with-zeros` I mentioned before.

Maybe make the API of `pad-with-zeros` take a keyword for prepend or postpend, doing the more useful one by default.

File: src/geb/geb.lisp [hunk 40]
```diff
@@ -538,3 +538,45 @@ I am the [GAPPLY][generic-function] for a generic [OPAQUE][class] I
...
+                     (error "Pairing ill-defined for ~A" morph))))
+    ((or substobj
+         init
+         terminal
+         project-left
+         project-right
+         inject-left
+         inject-right
+         distribute)
+     t)))
```
Comment (mariari) [inline src/geb/geb.lisp:576]: Yeah just newline the code after the case.

```lisp
(defmethod well-defp-cat ((morph <substmorph>))
  (etypecase-of substmorph morph
    (comp
     (let ((mcar (mcar morph))
           (mcadr (mcadr morph)))
       (if (and (well-defp-cat mcar)
                (well-defp-cat mcadr)
                (obj-equalp (dom mcar)
                            (codom mcadr)))
           t
           (error "(Co)Domains do not match for ~A" morph))))
    (case
        (let ((mcar (mcar morph))
              (mcadr (mcadr morph)))
          (if (and (well-defp-cat mcar)
                   (well-defp-cat mcadr)
                   (obj-equalp (codom mcar)
                               (codom mcadr)))
              t
              (error "Casing ill-defined for ~A" morph))))
    (pair
     (let ((mcar (mcar morph))
           (mcdr (mcdr morph)))
       (if (and (well-defp-cat mcar)
                (well-defp-cat mcdr)
                (obj-equalp (dom mcar)
                            (dom mcdr)))
           t
           (error "Pairing ill-defined for ~A" morph))))
    ((or substobj
         init
         terminal
         project-left
         project-right
         inject-left
         inject-right
         distribute)
     t)))
```

Comment (mariari) [reply src/geb/geb.lisp:576]: In the future, make such generic functions be a separate topic. Thus this should be apart of some other topic, and merged into your topic.

For this PR I'll let you keep it here, but note for the future, will make your work diverge less in a real setting of X engineers hacking on a project and if I was an active dev I could merge this patch into my own code

File: src/seqn/seqn.lisp [hunk 41]
```diff
@@ -195,3 +195,62 @@ is capable of succesfully evaluating all compiled terms"
...
+         inj-size
+         zero-bit
+         one-bit
+         seqn-add
+         seqn-multiply
+         seqn-divide
+         seqn-subtract
+         seqn-nat
+         seqn-concat
+         seqn-decompose
```
Comment (mariari) [inline src/seqn/seqn.lisp:253]: Same formatting comment as I mentioned earlier

File: test/seqn.lisp [hunk 42]
```diff
@@ -1 +1,23 @@
...
+  :parent geb-seqn
+  (is obj-equalp (list 1 0) (gapply (geb.seqn.spec:seqn-decompose 3) (list 4)))
+  (is obj-equalp (list 4) (gapply (geb.seqn.spec:seqn-concat 1 2) (list 1 0)))
+  (is obj-equalp (list 4) (gapply (geb.seqn.spec:composition (geb.seqn.spec:seqn-concat 1 2)
+                                                             (geb.seqn.spec:seqn-decompose 3))
+                                  (list 4)))
+  (is obj-equalp (list 1 0) (gapply (geb.seqn.spec:composition (geb.seqn.spec:seqn-decompose 3)
+                                                               (geb.seqn.spec:seqn-concat 1 2))
+                                  (list 1 0))))
+
```
Comment (mariari) [inline test/seqn.lisp:22]: alias your seqn so this is just seqn: and remove the seqn words on your functions they only make the API ugly

**Review by mariari — CHANGES_REQUESTED** [review]
Yeah see all my comments, they are fairly through.

In the future take the generic work into their own topics so I can merge them faster and be done more properly.

When changes are in big prs like this I get less stringent and works worse with a bigger team size.

Also the docs won't build this isn't your fault (I don't think besides warnings you produced), but by other issues I've posted in #140

**Review by mariari — CHANGES_REQUESTED** [review]
Yeah see all my comments, they are fairly through.

In the future take the generic work into their own topics so I can merge them faster and be done more properly.

When changes are in big prs like this I get less stringent and works worse with a bigger team size.

Also the docs won't build this isn't your fault (I don't think besides warnings you produced), but by other issues I've posted in #140

File: src/geb/trans.lisp (hunk 35 above)
Comment (agureev) [reply src/geb/trans.lisp:299]: With the constructors we currently have for the SeqN category this is the canonical way to do this. We can have extra constructors just skipping entries on left-right instead of only doing this for nil-bit entries.

File: src/vampir/vampir.lisp [hunk 43]
```diff
@@ -173,10 +172,237 @@
...
+  (make-alias
+   :name :take_base
+   :inputs (list :lst)
+   :body (list (make-wire :var :[]))))
+
+(defparameter *take-ind*
+  (make-alias
+   :name :take_ind
+   :inputs (list :take
+                 :|(h:t)|)
```
[the remark this answers is not in the document]
Comment (agureev) [reply src/vampir/vampir.lisp:372]: Right, after this commit we will probably do small patches for things like these

File: src/seqn/trans.lisp (hunk 24 above)
Comment (agureev) [reply src/seqn/trans.lisp:45]: This is a good point and I agree that mixing generation and optimization is messy, but I think this is something for another patch. I'll leave this untreated for now and address it after we release v.0.5.0 since I am not sure of the most efficient way here. Reducing everything as things generate makes it easier for internal printing to handle stuff.

File: src/seqn/trans.lisp (hunk 17 above)
Comment (agureev) [reply src/seqn/trans.lisp:260]: Good idea!

File: src/specs/seqn-printer.lisp (hunk 26 above)
Comment (agureev) [reply src/specs/seqn-printer.lisp:7]: Good idea

File: src/seqn/seqn.lisp [hunk 44]
```diff
@@ -0,0 +1,139 @@
...
+      (let ((lh1 (length l1))
+            (lh2 (length l2)))
+        (if (= lh1 1)
+            (list (list (car l1) (car l2)))
+            (append (prod-list (butlast l1 1)
+                               (butlast l2 1))
+                    (list
+                     (list (nth (1- lh1) l1) (nth (1- lh2) l2))))))))
+
+(defun seq-max-fill (seq1 seq2)
```
[the remark this answers is not in the document]
Comment (agureev) [reply src/seqn/seqn.lisp:32]: Something that I will address in another topic. I agree that the naming is weird.

File: src/seqn/seqn.lisp (hunk 2 above)
[the remark this answers is not in the document]
Comment (agureev) [reply src/seqn/seqn.lisp:42]: We'd need to unpack lists for it to work as intended. I'd rather keep it this way for now at least

File: src/seqn/trans.lisp (hunk 21 above)
Comment (agureev) [reply src/seqn/trans.lisp:354]: Maybe I'll make it a separate function

File: src/geb/geb.lisp (hunk 31 above)
Comment (agureev) [reply src/geb/geb.lisp:348]: I thought this was a typo. I'll remove this change
